Replace invalid hostname characters for task names and docker hostname

sbeaulie opened the pull request:
Before this change, if you used the mesos plugin option 'Docker Image Can Be Customized'
and the resulting task name was less than 64 characters it could happen that the docker
image name contained invalid characters both for the task name and the docker hostname.
For example a custom label:foo/bar:latest would fail as a task name because / is invalid
in a task name since it breaks the directory structure. It would also fail as a docker
hostname since it needs to comply with RFC 1123 hostnames

Original commit was to fix https://github.com/jenkinsci/mesos-plugin/issues/251

sbeaulie commented:
candidate fix for #295 too

eThomasHerrlin commented:
More RFC compliant: `name.replaceAll("[^a-zA-Z0-9-]", "-");`
"_" is not allowed and "." may cause problems also, safest to remove them.

RFC 952:
[...]alphabet (A-Z), digits (0-9), minus
   sign (-), and period (.).  Note that periods are only allowed when
   they serve to delimit components of "domain style names".[...]
Later RFCs extends length and removes limit of first char being Alpha.
Possibly also need to left strip any "-" after replace.

sbeaulie commented:
@eThomasHerrlin I have changed the regexp accordingly.

By the way this is still an issue with the latest 0.18.1 release, for instance when setting a custom image to `bitnami/ruby:latest`

I get the following error in the jenkins logs:
```
Status update: task value: "mesos-jenkins-a88b6ab710254e9c97218e932244cead-worker:bitnami/r"
 is in state TASK_ERROR with message ''mesos-jenkins-a88b6ab710254e9c97218e932244cead-worker:bitnami/r' contains invalid characters'
```

I've rebased the fix on top of 0.18.1 and released it for my use, but is there any way to get traction on merging this PR?

vinodkone commented on src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java:
```diff
@@ -243,10 +243,26 @@ private static String joinPaths(String prefix, String suffix) {
+   * @param name the task or hostname
+   * @return cleaned up hostname or task name where invalid chars are replaced by the "_" character
+   */
+  private static String replaceInvalidHostnameCharacters(String name) {
```
Instead of fixing the task name, why dont we fix the slave name instead? That would be change in one place (MesosUtils.buildNodeName()) which I think is cleaner.

vinodkone replied:
I'll be happy to merge this once comments get resolved. Thanks for fixing the issue and sorry it took so long to get feedback.

sbeaulie replied:
would changing the slave name ripple through to the task name?

vinodkone replied:
yes, they are the same.

jeschkies commented:
Ping @sbeaulie 🙂

sbeaulie commented:
opening new PR
